omc3/knob_extractor.py

```diff
@@ -0,0 +1,459 @@
+        When extraction was not possible, the value attribute of the respective KnobEntry is still None
+
+    """
+    LOGGER.info("---- EXTRACTING KNOBS -------------------------")
+    LOGGER.info(f"extracting knobs for {time}")
+    knobs = {}
+
+    LOGGER.info("---- KNOBS ------------------------------------")
+    for category in knob_categories:
+        for knob in KNOB_CATEGORIES.get(category, [category]):
+            knobs[knob] = knobs_dict[knob]
+
+            LOGGER.info(f"Looking for {knob:<34s} ")
+            knobkey = f"LhcStateTracker:{knob}:target"
+            knobvalue = ldb.get(knobkey, time.timestamp())  # use timestamp to preserve timezone info
+            if knobkey not in knobvalue:
+                LOGGER.warning(f"No value for {knob} found!")
+                continue
+
+            LOGGER.debug(f"Some value for {knob} extracted.")
+            timestamps, values = knobvalue[knobkey]
+            if len(values) == 0:
+                LOGGER.warning(f"No value for {knob} found for given time!")
+                continue
+
+            value = values[-1]
+            if not math.isfinite(value):
+                LOGGER.warning(f"Value for {knob} is not a number or infinite!")
+                continue
+
+            LOGGER.info(f"Knob value for {knob} extracted: {value} (unscaled)")
+            knobs[knob].value = value
+    return knobs
+
+
+def _write_knobsfile(output: Union[Path, str], collected_knobs: KnobsDict, time):
+    """ Takes the collected knobs and writes them out into a text-file. """
+    collected_knobs = collected_knobs.copy()  # to not modify the return dict
+
+    # Sort the knobs by category
+    category_knobs = {c: {} for c in KNOB_CATEGORIES.keys()}
+    for category, names in KNOB_CATEGORIES.items():
+        for name in names:
+            if name in collected_knobs.keys():
+                category_knobs[category][name] = collected_knobs.pop(name)
+    category_knobs["Other Knobs"] = collected_knobs
+
+    # Write them out
+    with open(output, "w") as outfile:
+        outfile.write(f"!! --- knobs extracted by knob_extractor\n")
+        outfile.write(f"!! --- extracted knobs for time {time}\n\n")
+        for category, knobs in category_knobs.items():
+            if not knobs:
+                continue
+            outfile.write(f"!! --- {category:10} --------------------\n")
+            for knob, knob_entry in knobs.items():
+                outfile.write(f"{knob_entry.get_madx()}\n")
+            outfile.write("\n")
+        outfile.write("\n")
+
+
+# Knobs Dict -------------------------------------------------------------------
+
+def _get_knobs_def_file(user_defined: Optional[Union[Path, str]] = None) -> Path:
+    """ Check which knobs-definition file is appropriate to take. """
+    if user_defined is not None:
+        LOGGER.info(f"Using user defined knobs.txt: '{user_defined}")
+        return Path(user_defined)
+
+    if KNOBS_TXT_MDLDIR.is_file():
```

Comment:
See, here the variable with `MDLDIR` makes me think it's a directory, and we check for `.is_file()`. No me gusta :/